code/asteroid/asteroid.cpp

```diff
@@ -2195,107 +2273,190 @@ to do with the debris of ships that explode, however these
-		}
 
 		if (optional_string("$Briefing Icon Closeup Zoom:")) {
 			stuff_float(&Asteroid_icon_closeup_zoom);
 		}
-		else {
-			Asteroid_icon_closeup_zoom = 0.5f;	// magic number from retail
-		}
+
 	}
 	catch (const parse::ParseException& e)
 	{
-		mprintf(("TABLES: Unable to parse '%s'!  Error message = %s.\n", "asteroid.tbl", e.what()));
+		mprintf(("TABLES: Unable to parse '%s'!  Error message = %s.\n", filename, e.what()));
 		return;
 	}
 }
 
+// Gets the index of an asteroid by name from Asteroid_info
+static int get_asteroid_index(const char* asteroid_name)
+{
+	for (int i = 0; i < (int)Asteroid_info.size(); i++) {
+		if (!stricmp(asteroid_name, Asteroid_info[i].name)) {
+			return i;
+		}
+	}
+
+	return -1;
+}
+
+static void verify_asteroid_splits() 
+{
+	
+	for (int i = 0; i < (int)Asteroid_info.size(); i++) {
+		SCP_vector<int> splits_to_remove;
+
+		for (int j = 0; j < (int)Asteroid_info[i].split_info.size(); j++) {
+			int asteroid_index = get_asteroid_index(Asteroid_info[i].split_info[j].name);
+			if (asteroid_index >= 0) {
+				Asteroid_info[i].split_info[j].asteroid_type = asteroid_index;
+			} else {
+				Warning(LOCATION,
+					"Unknown asteroid %s defined as split type for asteroid %s, removing!",
+					Asteroid_info[i].split_info[j].name,
+					Asteroid_info[i].name);
+				splits_to_remove.push_back(j); // Remove them all after we're done so we don't mess with the loop
+			}
+		}
+
+		// Remove errored splits
+		for (int j = 0; j < (int)splits_to_remove.size(); j++) {
+			Asteroid_info[i].split_info.erase(Asteroid_info[i].split_info.begin() + splits_to_remove[j]);
+		}
+	}
+}
+
+static void verify_asteroid_list()
+{
+	SCP_string asteroid_size[NUM_ASTEROID_SIZES] = {"Small", "Medium", "Large"};
+	float asteroid_spawn_weight[NUM_ASTEROID_SIZES] = {SMALL_DEBRIS_WEIGHT, MEDIUM_DEBRIS_WEIGHT, LARGE_DEBRIS_WEIGHT};
+	
+	// Set the three asteroid sizes
+	for (int i = 0; i < NUM_ASTEROID_SIZES; i++) {
+
+		bool found = false;
+
+		for (int j = 0; j < (int)asteroid_list.size(); j++) {
+			if (asteroid_list[j].type == i) {
+				mprintf(("%s asteroid parsed with name %s\n", asteroid_size[i].c_str(), asteroid_list[j].name));
+				asteroid_list[j].spawn_weight = asteroid_spawn_weight[i]; //Be absolutely sure we're using the right spawn weight for asteroids
+				Asteroid_info.push_back(asteroid_list[j]);
+				asteroid_list.erase(asteroid_list.begin() + j);
+				found = true;
+				break;
+			}
+		}
+
+		if (found)
+			continue;
+
+		//Left this as a log print instead of a Warning because of retail-Mjn
+		mprintf(("%s asteroid not found. Using asteroid %s\n", asteroid_size[i].c_str(), asteroid_list[0].name));
+		asteroid_list[0].type = i;
+		asteroid_list[0].spawn_weight = asteroid_spawn_weight[i]; //Be absolutely sure we're using the right spawn weight for asteroids
+		Asteroid_info.push_back(asteroid_list[0]);
+		asteroid_list.erase(asteroid_list.begin());
```

Comment:
So yeah, same issue on this loop.  Erasing while looping through the vector will put you in a different place than you expect.  Because you have to use valid entries in asteroid_list in the next loop and then clear it, I suggest recording the *valid* indexes in this loop so that you can avoid bad entries in the next loop.  It also ends up being faster than erasing because erasing is relatively expensive (though we are not likely to see the difference unless there's a truly massive table).

Reply:
hmm. Actually I think this one is OK because we're not actually looping through the vector. We search `asteroid_list` for the small asteroid defined by type. If found it is added to `Asteroid_info` and removed from `asteroid_list`. Then we break, increment `i` and start over with `asteroid_list`.

Similarly if the small asteroid is not found defined by type we just take the first entry in `asteroid_list` instead of a particular one and do the same thing.

Help me see it if I'm missing the error.

Reply:
You're right. I didn't realize the break statement was also basically resetting that loop, so you're right, it doesn't matter that you deleted while iterating.